pkg/watch/manager.go

```diff
@@ -1,362 +1,380 @@
+	wm.watchedMux.Lock()
+	defer wm.watchedMux.Unlock()
+	return wm.doAddWatch(r, gvk)
+}
 
-	newWatchedKinds := make(map[schema.GroupVersionKind]vitals)
-	for gvk, vitals := range wm.watchedKinds {
-		if _, ok := removed[gvk]; !ok {
-			if newVitals, ok := changed[gvk]; ok {
-				newWatchedKinds[gvk] = newVitals
-			} else {
-				newWatchedKinds[gvk] = vitals
-			}
-		}
-	}
+func (wm *Manager) doAddWatch(r *Registrar, gvk schema.GroupVersionKind) error {
+	// lock acquired by caller
 
-	filteredNewWatchedKinds, err := wm.filterPendingResources(newWatchedKinds)
-	if err != nil {
-		return false, errp.Wrap(err, "could not filter new watched kinds, not restarting watch manager")
+	if r == nil {
+		return fmt.Errorf("nil registrar cannot watch")
 	}
-	if len(filteredNewWatchedKinds) != len(newWatchedKinds) {
-		var missing []string
-		for k := range newWatchedKinds {
-			if _, ok := filteredNewWatchedKinds[k]; !ok {
-				missing = append(missing, k.String())
-			}
-		}
-		log.Info("previously watched resources have gone pending, removing them from the watch list", "pending", missing)
+
+	// watchers is everyone who is *already* watching.
+	watchers := wm.watchedKinds[gvk]
+
+	// m is everyone who *wants* to watch.
+	m := wm.managedKinds.Get() // Not a deadlock but beware if assumptions change...
+	if _, ok := m[gvk]; !ok {
+		return fmt.Errorf("could not mark %+v as managed", gvk)
 	}
 
-	for gvk, vitals := range readyToAdd {
-		filteredNewWatchedKinds[gvk] = vitals
+	// Sanity
+	if !m[gvk].registrars[r] {
+		return fmt.Errorf("registrar %s not in desired watch set", r.parentName)
 	}
 
-	if err := wm.restartManager(filteredNewWatchedKinds); err != nil {
-		return false, errp.Wrap(err, "could not restart watch manager: %s")
+	if watchers.registrars[r] {
+		// Already watching.
+		return nil
 	}
 
-	wm.watchedKinds = filteredNewWatchedKinds
-	return true, nil
-}
+	u := &unstructured.Unstructured{}
+	u.SetGroupVersionKind(gvk)
+	informer, err := wm.cache.GetInformerNonBlocking(u)
+	if err != nil || informer == nil {
+		// This is expected to fail if a CRD is unregistered.
+		return fmt.Errorf("getting informer for kind: %+v %w", gvk, err)
+	}
 
-// Start looks for changes to the watch roster every 5 seconds. This method has a
-// benefit compared to restarting the manager every time a controller changes the watch
-// of placing an upper bound on how often the manager restarts.
-func (wm *Manager) Start(done <-chan struct{}) error {
-	ticker := time.NewTicker(5 * time.Second)
-	defer ticker.Stop()
-	for {
-		select {
-		case <-done:
-			log.Info("watch manager shutting down")
-			wm.close()
-			return nil
-		case <-ticker.C:
-			if _, err := wm.updateOrPause(); err != nil {
-				log.Error(err, "error in updateManagerLoop")
-			}
+	switch {
+	case len(watchers.registrars) > 0:
+		// Someone else was watching, replay events in the cache to the new watcher.
+		if err := wm.replayEvents(r, gvk); err != nil {
```

Comment:
`replayEvents()` appears to be a blocking call that will wait for the informer to warm its cache due to the call to `List()`, or am I missing something?

Could we run this in a separate goroutine? We would need some sort of context that we could cancel if the resource became unwatched while the list was still replaying. Or we could just rely on the receiver to filter out things it's not interested in, like we are doing with the sync controller. The context solution seems better as it avoids an explosion of goroutines if there is a flapping watch/no-watch cycle.

Flapping is a risk we should consider, some kind of debouncing may be necessary to avoid repeated List() calls.

Reply:
I've implemented something along these lines in the latest push, please take a look!
Goroutines are contained - there will only be a single per-registrar-per-gvk while replaying events.

Regarding flapping: List calls will be against the informer cache, not the API Server. Assuming there is another registrar watching the same gvk (which is necessary to trigger the replay logic) - a second registrar repeatedly unregistering and registering should not be an issue.

If we are concerned against the primary (first) registrar unregistering and re-registering, we could add rate limiting in the future I'd propose considering that as followup work.